tripy/tripy/backend/compiler_api.py

```diff
@@ -31,6 +31,169 @@
 from tripy.common.shape_bounds import ShapeBounds
 from tripy.frontend import Tensor, Trace
 from tripy.utils import json as json_utils
+from tripy.common.device import device as Device
+
+
+# Global variable to store instances
+_default_stream_instances = {}
+
+
+@export.public_api(document_under="compiler")
+def default_stream(device: Device = Device("gpu")) -> "Stream":
+    """
+    Provides access to the default CUDA stream for a given device.
+    This function implements singleton pattern to ensure a single default stream instance per device.
+
+    Args:
+        device (Device, optional): The device for which to get the default stream. Defaults to Device("gpu").
+
+    Returns:
+        Stream: The default Stream object for the specified device.
+
+    Raises:
+        ValueError: If the device is not of type 'gpu' or if the device index is not 0.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Use streams in execution
+
+        default_stream = tp.default_stream()
+        linear = tp.Linear(2, 3)
+        compiler = tp.Compiler(linear)
+
+        compiled_linear = compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32))
+
+        a = tp.ones((2, 2), dtype=tp.float32)
+
+        compiled_linear.stream = default_stream
+        out = compiled_linear(a)
+        default_stream.synchronize()
+        assert cp.array_equal(cp.from_dlpack(out), cp.from_dlpack(linear(a)))
+
+    Usage:
+        default = default_stream()  # Returns the default Stream object for the current device
+        default_gpu1 = default_stream(tp.device("gpu", 1))  # Returns the default Stream object for GPU 1
+
+    Note:
+        Calling default_stream() with the same device always returns the same Stream instance for that device.
+    """
+    global _default_stream_instances
+
+    if device.kind != "gpu":
+        raise ValueError(f"default_stream creation requires device to be of type gpu, got device={device}.")
+
+    if device.index != 0:
+        raise ValueError(f"Tripy stream only works with device index 0, got device={device}")
+
+    if device.index not in _default_stream_instances:
+        _default_stream_instances[device.index] = Stream()
+
+    return _default_stream_instances[device.index]
+
+
+@export.public_api(document_under="compiler")
+def synchronize() -> None:
+    """
+    Synchronizes the default stream.
```

Comment:
I think this should synchronize all streams (that's the behavior in torch)? Otherwise it's not really adding any value - we can just do `default_stream().synchronize()`. 
Also, this won't work for multi-device. Maybe just remove it for now?

Reply:
Removing for now but I think there is value in it (will file issue to bring it back).